sycl/include/sycl/ext/oneapi/properties/properties.hpp

```diff
@@ -9,265 +9,180 @@
+// base case is only for invalid cases to suppress compiler errors about base
+// class storage of properties
+template <class L, class = void> class properties {
+  static_assert(detail::mp11::mp_all_of<L, std::is_class>(),
                 "Unrecognized property in property list.");
-  static_assert(detail::IsSorted<PropertiesT>::value,
-                "Properties in property list are not sorted.");
-  static_assert(detail::SortedAllUnique<PropertiesT>::value,
+  static_assert(detail::mp11::mp_is_set<L>(),
                 "Duplicate properties in property list.");
-  static_assert(detail::NoConflictingProperties<PropertiesT>::value,
-                "Conflicting properties in property list.");
-
-public:
-  template <typename... PropertyValueTs>
-  constexpr properties(PropertyValueTs... props)
-      : Storage(detail::ExtractProperties<StorageT>::Extract(
-            std::tuple<PropertyValueTs...>{props...})) {}
+};
 
-  template <typename PropertyT>
-  static constexpr std::enable_if_t<detail::IsProperty<PropertyT>::value, bool>
-  has_property() {
-    return detail::ContainsProperty<PropertyT, PropertiesT>::value;
-  }
+template <template <class...> class TL, class... V>
+class properties<TL<V...>,
+                 std::enable_if_t<(std::is_class_v<V> && ...) &&
+                                  detail::mp11::mp_is_set<TL<V...>>::value>>
+    : V... {
+private:
+  using map = detail::property_map<properties>; // map of properties
+  static_assert(detail::mp11::mp_is_map<map>(),
+                "Duplicate properties in property list.");
+  using keys = detail::mp11::mp_map_keys<map>;
+  using conflicting = detail::mp11::mp_apply<
+      detail::mp11::mp_append,
+      detail::mp11::mp_transform<detail::ConflictingProperties, keys>>;
+  static_assert(
+      !detail::mp11::mp_any_of_q<
+          conflicting,
+          detail::mp11::mp_bind_front<detail::mp11::mp_set_contains, keys>>(),
+      "Conflicting properties in property list.");
 
-  template <typename PropertyT>
-  typename std::enable_if_t<detail::IsRuntimeProperty<PropertyT>::value &&
-                                has_property<PropertyT>(),
-                            PropertyT>
-  get_property() const {
-    return std::get<PropertyT>(Storage);
+public:
+  template <class... T>
+  constexpr properties(T... v)
+      : T(v)... {} // T might have different ordering than V
+  template <class P> static constexpr bool has_property() {
+    return detail::mp11::mp_map_contains<map, P>();
   }
-
-  template <typename PropertyT>
-  typename std::enable_if_t<detail::IsRuntimeProperty<PropertyT>::value &&
-                                !has_property<PropertyT>(),
-                            void>
-  get_property() const {
-    static_assert(has_property<PropertyT>(),
+  template <class P> static constexpr auto get_property(int = 0) {
+    using T = detail::mp11::mp_map_find<map, P>;
+    static_assert(!std::is_same_v<T, void>,
```

Comment:
Is this actually faster than just using `has_property`?

Reply:
propably not. `has_property` calls `mp_map_contains` which calls `mp_map_find` and compares it against `void`.  And the compiler should memorize the `mp_map_find` template instanstiation and therefore it shouldn't matter if it is called twice.